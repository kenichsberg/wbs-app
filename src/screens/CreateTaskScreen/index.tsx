import * as React from 'react';
import 'react-native-gesture-handler';
import { Ionicons } from '@expo/vector-icons';
import { Container, Header, Content, Text, View, Body, Right, Button, Form, Item, Input, Label, Picker, Icon, DatePicker, List, ListItem } from 'native-base';
import DateTimePickerModal from "react-native-modal-datetime-picker";
import { CreateTaskProps } from '/navigations/types.tsx';

const moment = require('moment');

export type Task = {
  id: number;
  category: string;
  taskName: string;
  startDatetimePlanned: string;
  endDatetimePlanned: string;
  //startDatetimeResult: string;
  //endDatetimeResult: string;
  selectedDocument: number;
};

type DatePickerVisibilities = {
  startDatetimePlanned: boolean;
  endDatetimePlanned: boolean;
  //startDatetimeResult: boolean;
  //endDatetimeResult: boolean;
};


export const CreateTaskScreen: React.FC<CreateTaskProps> = ({ navigation, route }) => {
  // id
  const [id, setId] = React.useState<number | null>(null);
  // 分類
  const [category, setCategory] = React.useState<string>('');
  // プロセス名
  const [taskName, setTaskName] = React.useState<string>('');

  // datetime初期値を設定
  const now = new Date();
  now.setMinutes(0);
  now.setSeconds(0);
  // 予定開始日時
  const [startDatetimePlanned, setStartDatetimePlanned] = React.useState<Date>(now);
  // 予定終了日時
  const [endDatetimePlanned, setEndDatetimePlanned] = React.useState<Date>(now);
  // 実績開始日時
  const [startDatetimeResult, setStartDatetimeResult] = React.useState<Date | null>(null);
  // 実績終了日時
  const [endDatetimeResult, setEndDatetimeResult] = React.useState<Date | null>(null);
  // 成果物
  const [selectedDocument, setSelectedDocument] = React.useState<number | null>(null);


  // 日時選択モーダル　表示・非表示
  const [datePickerVisibilities, setDatePickerVisibilities] = React.useState<DatePickerVisibilities>({
    startDatetimePlanned: false,
    endDatetimePlanned: false,
    //startDatetimeResult: false,
    //endDatetimeResult: false,
  });

  const task: Task = {
    id: id,
    category: category,
    taskName: taskName,
    startDatetimePlanned: JSON.stringify(startDatetimePlanned),
    manHour: parseFloat(manHour),
    endDatetimePlanned: JSON.stringify(endDatetimePlanned),
    /*
    startDatetimeResult: startDatetimeResult
      ? JSON.stringify(startDatetimeResult)
      : 'null',
    endDatetimeResult: endDatetimeResult
      ? JSON.stringify(endDatetimeResult)
      : 'null',
     */
    selectedDocument: selectedDocument
  };


  React.useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Button 
          data-test="create-button"
          hasText
          transparent
          onPress={ (): void => {
            navigation.navigate('TaskList', { task: task });
          }}
          style={{ 
            marginRight: 20,
          }}
        >
          <View style={{ 
              flexDirection: 'row',
              justifyContent: "space-between",
              alignItems: "center"
            }}
          >
            <Icon 
              name="checkmark" 
              style={{ 
                color: '#1F5E56' ,
                marginRight: 5,
              }} 
            />
            <Text style={{ 
                color: '#1F5E56',
                fontWeight: "600",
              }}
            >
               作成
            </Text>
          </View>
        </Button>
      ),
    });
  }, [navigation]);


  // 引数を受け取った時の処理
  React.useEffect(() => {
    if (route.params?.task) {
      const values: Task = route.params.task;

      setId(values.id);
      setCategory(values.category);
      setTaskName(values.taskName);
      setStartDatetimePlanned(new Date(JSON.parse(values.startDatetimePlanned)));
      setEndDatetimePlanned(new Date(JSON.parse(values.endDatetimePlanned)));

      /*
      const startDatetimeResult = JSON.parse(values.startDatetimeResult) == null
        ? now
        : new Date(JSON.parse(values.startDatetimeResult));
      setStartDatetimeResult(startDatetimeResult);

      const endDatetimeResult = JSON.parse(values.endDatetimeResult) == null
        ? now
        : new Date(JSON.parse(values.endDatetimeResult));
       */
      setEndDatetimeResult(endDatetimeResult);

      setSelectedDocument(values.selectedDocument);

    }
  }, [route.params?.task]);


  // 日時選択モーダルの表示・非表示切り替え
  const handleDatePicker = (shouldShow: boolean, name: string): void => {
    setDatePickerVisibilities({ ...datePickerVisibilities, [name]: shouldShow });
  };

  // map of name -> setState()
  const nameToSetStateFunc = new Map([
    ['startDatetimePlanned', setStartDatetimePlanned],
    ['endDatetimePlanned', setEndDatetimePlanned],
    ['startDatetimeResult', setStartDatetimeResult],
    ['endDatetimeResult', setEndDatetimeResult]
  ]);

  // 日時選択モーダル　選択時の処理
  const handleConfirm = (date: Date, name: string): void => {
    const func = nameToSetStateFunc.get(name);
    func(date);

    handleDatePicker(false, name);
  };

  // datetime型入力フィールドを取得
  const getDatetimeInputField = (value: Date, name: string, label: string): JSX.Element => {
    return (
      <Item stackedLabel>
        <Label>{ label }</Label>
        <DateTimePickerModal
          isVisible={ datePickerVisibilities[name] }
          mode="datetime"
          date={ value }
          onConfirm={ event => handleConfirm(event, name) }
          onCancel={ () => handleDatePicker(false, name) }
          headerTextIOS="選択してください"
          confirmTextIOS="OK"
          cancelTextIOS="キャンセル"
        />
        <ListItem onPress={ () => handleDatePicker(true, name) } >
          <Text>
            { moment(value).format('YYYY年MM月DD日 HH:mm') }
          </Text>
          <Button 
            small
            primary
            transparent
            onPress={ () => handleDatePicker(true, name) }
          >
            <Text>変更する</Text>
          </Button>
        </ListItem>
      </Item>
    );
  }

  // JSX
  return (
    <>
      <Container>
        <Content padder>
          <Form>
            <Item stackedLabel>
              <Label>分類</Label>
              <Input
                value={ category }
                onChangeText={ setCategory }
              />
            </Item>
            <Item stackedLabel>
              <Label>プロセス名</Label>
              <Input
                value={ taskName }
                onChangeText={ setTaskName }
              />
            </Item>

            { getDatetimeInputField(startDatetimePlanned, 'startDatetimePlanned', '予定開始日時') }
            { getDatetimeInputField(endDatetimePlanned, 'endDatetimePlanned', '予定終了日時') }
            {/*
            { route.params?.task == null
                ? null
                : getDatetimeInputField(startDatetimeResult, 'startDatetimeResult', '実績開始日時') }
            { route.params?.task == null
                ? null
                : getDatetimeInputField(endDatetimeResult, 'endDatetimeResult', '実績終了日時') }
              */}

            <Item stackedLabel picker last>
              <Label>成果物</Label>
              <Picker
                mode="dropdown"
                iosIcon={ <Icon name="arrow-down" /> }
                style={{ width: undefined }}
                placeholder="選択してください"
                placeholderStyle={{ color: "#bfc6ea" }}
                placeholderIconColor="#007aff"
                selectedValue={ selectedDocument }
                onValueChange={ event => setSelectedDocument(event) }
              >
                <Picker.Item label="設計書" value="0" />
                <Picker.Item label="単体検査項目" value="1" />
                <Picker.Item label="開発" value="2" />
                <Picker.Item label="総合テスト実績" value="3" />
                <Picker.Item label="移行チェックシート" value="4" />
              </Picker>
            </Item>
          </Form>
        </Content>
        {/*
        <Button
          data-test="create-button"
          block
          rounded
          onPress={ () => {
            navigation.navigate('TaskList', { task: task });
          }}
          style={{ 
            marginTop: 20,
            marginBottom: 10,
            marginLeft: 10,
            marginRight: 10,
            backgroundColor: '#1F5E56'
          }}
        >
          <Text style={{ fontWeight: "600" }}>作成する</Text>
        </Button>
         */}
      </Container>
    </>
  );
}
