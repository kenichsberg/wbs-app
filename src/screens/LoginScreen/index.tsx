import * as React from 'react';
import { AsyncStorage } from 'react-native';
import { View, Text, Button } from 'native-base';
import { getGithubToken } from '/components/Login/';
import { firebase } from '/data-access/firebase';
import { UserCredential } from '@firebase/auth-types/';


export const LoginScreen: React.FC = () => {

const loginWithGithub = async (param: undefined | string)
  : Promise<undefined | UserCredential> => 
{
  try {
    const token = param ??  await getGithubToken();
    console.log('token: ', token);

    if (!token) return undefined;

    AsyncStorage.setItem('githubToken', token);

    const credential = await firebase.auth.GithubAuthProvider.credential(token);
    console.log('credential: ', credential);

    return firebase.auth().signInWithCredential(credential);

  } catch (error) {
    console.log(error);
    return undefined;

  }
};

  return (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <Text>ログイン画面</Text>
      <Button
        onPress={ () => loginWithGithub(undefined) }
      >
        <Text>GitHubログイン</Text>
      </Button>
    </View>
  );
}
