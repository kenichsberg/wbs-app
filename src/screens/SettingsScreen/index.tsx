import * as React from 'react';
import { View, Text, Button } from 'native-base';

export const SettingsScreen: React.FC = ({ navigation }) => {
  return (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <Text>設定画面</Text>
      <Button
        title="Go to Details"
        onPress={() => navigation.navigate('Details')}
      />
    </View>
  );
}
