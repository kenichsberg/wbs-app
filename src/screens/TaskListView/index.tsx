import * as React from 'react';
import 'react-native-gesture-handler';
import { AppStateContext } from '/contexts/AppStateContext';
import { ScrollView } from 'react-native';
import { View, Body, Right, Text, Button, List, ListItem, Icon } from 'native-base';
import { getFormattedTasks } from '/domain/Task/';
import { Task } from '/domain/Task/';
import { ListTabProps } from '/navigations/types.tsx';
import { parseJsonToMoment } from '/services/Date/'; 
import { Color } from '/style/Color';


type Props = 
  {
    navigation: ListTabProps['navigation']
  }
  & {
    tasks: Array<Task>;
  };

type ListItemProps = {
  item: Task,
  index: number,
  navigation: ListTabProps['navigation']
};


const getDateString = (jsonDate = '"null"'): string => {
  if (jsonDate === '"null"') {
    return '-';
  }

  const date = parseJsonToMoment(jsonDate);
  
  return date.format('YYYY/MM/DD HH:mm');
};



//export const TaskListView: React.FC<ListTabProps> = ({ navigation, route }) => {
export const TaskListView: React.FC<Props> = ({ navigation }) => {

  //const { tasks } = route.params;
  const { tasks } = React.useContext(AppStateContext);
  console.log(tasks);

  const { categories, tasksFormatted } = getFormattedTasks(tasks);

  // リスト1つのJSXを取得
  const getTaskList: React.FC<ListItemProps> = ({ item, navigation }) => {
    return (
        <ListItem key={ item.id }>
          <Body>
            <Text>{ item.taskName }</Text>
            <View style={{ marginLeft: 20, marginTop: 5 }}>
              <Text note>
                Start：{ getDateString(item.startDatetimePlanned) }
              </Text>
              <Text note>
                End  ：{ getDateString(item.endDatetimePlanned) }
              </Text>
              {/*
              <Text note>成果物: { item.selectedDocument }</Text>
                */}
            </View>
          </Body>
          <Right>
            <Button
              transparent
              data-test="edit-button"
              onPress={ () => navigation.navigate('EditTask', { task: item }) }
            >
              <Icon 
                name="ios-create" 
                style={{ color: Color.semiLight }}
              />
            </Button>
          </Right>
        </ListItem>
    );
  };

  return (
    <ScrollView>
      {
        tasks.length  === 0
          ? <ListItem>
              <Text >タスクを追加してください</Text>
            </ListItem>
          : categories.map(category => {
              return (
                <List key={ category }>
                  <ListItem itemDivider>
                    <Text>{ category }</Text>
                  </ListItem>
                  { 
                    tasksFormatted[category].map((item, index) => {
                      return (
                        getTaskList({ item, index , navigation})
                      );
                    })
                  }
                </List>
              )
            })
      }
    </ScrollView>
  );

}
