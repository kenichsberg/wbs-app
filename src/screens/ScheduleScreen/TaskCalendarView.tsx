import * as React from 'react';
import { Dimensions } from 'react-native';
import 'react-native-gesture-handler';
import { Container, Segment, Content, View, Body, Right, Button, List, ListItem, Separator, Icon, Fab } from 'native-base';
import moment from 'moment';
import Svg, { Line, Text } from 'react-native-svg';
import { FormatTasks } from '/components/FormatTasks';
import { createMaterialTopTabNavigator } from '@react-navigation/material-top-tabs';
import { GanttChart } from '/components/GanttChart';
import { TaskListProps } from '/navigations/types.tsx';
import { Task } from '/screens/CreateTaskScreen';

type Props = TaskListProps & Task;

const TopTab = createMaterialTopTabNavigator();

export const TaskCalendarView: React.FC<Props> = props => {
  // 引数
  const {tasks} = props;

  const { categories, tasksFormatted } = FormatTasks(tasks);

  return (
    <GanttChart tasks={tasks} />
  );
};

  /*
    <TopTab.Navigator>
      <TopTab.Screen name="Home" component={HomeScreen} />
      <TopTab.Screen name="Settings" component={SettingsScreen} />
    </TopTab.Navigator>
    */
/*
// 定数
const windowWidth = Dimensions.get('window').width;
const windowHeight = Dimensions.get('window').height;

const paddingX = 20;
const paddingY = 30;

const chartWidth = windowWidth - paddingX * 2;
const chartHeight = windowHeight - paddingY * 2;

const dayWidth = chartWidth / 6;

const factor = 1000 * 60 * 60 * 24;

const getTaskLength = (diffMilliseconds: int): int => {
  const diffDays = diffMilliseconds / factor;
  const int = String(diffDays).split('.')[0];
  const decimal = diffDays - int;
  const ratio = Math.floor(decimal * 24 / 8);

  return int * dayWidth + ratio * dayWidth;
}

export default function TaskCalendarlView(props) {
  // 引数
  const {tasks} = props;

  const [categories, tasksFormatted] = FormatTasks(tasks);

  const getGantt = (item, index) => {
    const origin = new Date(2020, 4, 5, 9, 0);
    const legthPlanned = getTaskLength(item.taskLength * 60 * 60 * 1000);
    const xStartPlanned = (item.startDatetimePlanned - origin) / factor * dayWidth;
    const xEndPlanned = xStartPlanned + legthPlanned;
    const xStartResult = (item.startDatetimeResult - origin) / factor * dayWidth;
    const lengthResult = (item.endDatetimeResult - item.startDatetimeResult) / factor * dayWidth;
    const xEndResult = xStartResult + lengthResult;
    const yTop = 30 + 40 * index;
    console.log(item.startDatetimePlanned, origin, item.startDatetimePlanned - origin, xStartPlanned);
    return (
      <>
        <Text x={xStartPlanned} y={yTop} fill="white">{item.taskName}</Text>
        <Line x1={xStartPlanned} y1={yTop + 10} x2={xEndPlanned} y2={yTop + 10} stroke="white" strokeWidth="2" />
        <Line x1={xStartResult} y1={yTop + 20} x2={xEndResult} y2={yTop + 20} stroke="white" strokeWidth="3" />
      </>
    );
  };

  const item ={
    taskName: 'sample Gantt',
    startDatetimePlanned: new Date(2020, 4, 5, 9, 0),
    endDatetimePlanned: new Date(2020, 4, 5, 21, 0),
    startDatetimeResult: new Date(2020, 4, 5, 9, 0),
    endDatetimeResult: new Date(2020, 4, 6, 11, 0),
    taskLength: 12
  };

  // JSX
  return (
    <>
      <View style={{backgroundColor: '#051B26', paddingLeft: paddingX, paddingRight: paddingX, paddingTop: paddingY, paddingButtom: paddingY }}>
        <Svg height="1300" width={windowWidth}>

          {[...Array(7).keys()].map(i => (
            <Line 
              key={i}
              x1={dayWidth * i} 
              y1="0" 
              x2={dayWidth * i} 
              y2={chartHeight} 
              stroke="#3E6B57" 
              strokeWidth="1" 
            />
          ))}

          {getGantt(item, 0)}
        </Svg>
      </View>
    </>
  );

}
*/
