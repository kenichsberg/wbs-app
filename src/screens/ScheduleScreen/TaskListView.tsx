import * as React from 'react';
import 'react-native-gesture-handler';
import { Container, Segment, Content, View, Body, Right, Text, Button, List, ListItem, Separator, Icon, Fab } from 'native-base';
import moment from 'moment';
import { FormatTasks } from '/components/FormatTasks'
import { TaskListProps } from '/navigations/types.tsx';
import { Task } from '/screens/CreateTaskScreen';

type Props = TaskListProps & Task;


// 日付（期間）の文字列を取得
const getPeriodString = (jsonDateStart: string, jsonDateEnd: string): string => {
  const dateStart = JSON.parse(jsonDateStart);
  const dateEnd = JSON.parse(jsonDateEnd);

  if (dateStart == null && dateEnd == null) {
    return '未入力';
  }

  const dateStartString = dateStart == null 
    ? ''
    : moment(JSON.parse(jsonDateStart)).format('YYYY年MM月DD日 HH:mm');

  const dateEndString = dateEnd == null 
    ? '対応中'
    : moment(JSON.parse(jsonDateEnd)).format('YYYY年MM月DD日 HH:mm');

  return `${dateStartString} 〜 ${dateEndString}`;

}



export const TaskListView: React.FC<Props> = props => {
  // 引数
  const {tasks, navigation} = props;

  const { categories, tasksFormatted } = FormatTasks(tasks);

  // リスト1つのJSXを取得
  const getTaskList = (item, index) => {
    return (
        <ListItem key={`ListItem-${index}`}>
          <Body>
            <Text>{item.taskName}</Text>
            <Text note>予定：{getPeriodString(item.startDatetimePlanned, item.endDatetimePlanned)}</Text>
            <Text note>実績：{getPeriodString(item.startDatetimeResult, item.endDatetimeResult)}</Text>
            <Text note>成果物: {item.selectedDocument}</Text>
          </Body>
          <Right>
            <Button transparent>
              <Icon 
                name="ios-create" 
                style={{color: 'tomato'}}
                onPress={() => navigation.navigate('EditTask', {task: item })}
              />
            </Button>
          </Right>
        </ListItem>
    );
  };

  // JSX
  return (
    <>
      {
        tasks.length  === 0
          ? <ListItem>
              <Text >タスクを追加してください</Text>
            </ListItem>
          : categories.map (category => {
              return (
                <List key={category}>
                  <Separator bordered key={`Separator-${category}`}>
                    <Text>{category}</Text>
                  </Separator>
                  {tasksFormatted[category].map ((item, index) => getTaskList(item, index))}
                </List>
              )
            })
      }
    </>
  );

}
