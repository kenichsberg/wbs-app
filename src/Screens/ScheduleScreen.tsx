import * as React from 'react';
import { StyleSheet, TextInput } from 'react-native';
import 'react-native-gesture-handler';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import * as Font from 'expo-font';
import { Ionicons } from '@expo/vector-icons';
import { Container, Segment, Content, View, Body, Right, Text, Button, List, ListItem, Separator, Icon, Fab } from 'native-base';
import moment from 'moment';
import TaskListView from '@Screens/TaskListView';
import TaskCalendarView from '@Screens/TaskCalendarView';


export default function ScheduleScreen({ navigation, route }) {

  // プロセスオブジェクトを保持するstate
  const [tasks, setTasks] = React.useState([]);

  // ビュー切り替え用state
  const [viewType, setViewType] = React.useState('LIST');

  // 引数を格納する変数
  let param = {};

  // 引数を受け取った時の処理
  React.useEffect(() => {
    if (route.params?.task) {
      param = route.params.task;

      // idがない場合は採番
      param.id = param.id ?? tasks.length;
      // paramと同じidがある場合は取り除く
      const tasksUnique = tasks.filter(task => task.id !== param.id);
      
      setTasks([...tasksUnique, param]);
    }
  }, [route.params?.task]);

  // リストビュー
  const taskList = <TaskListView tasks={tasks} navigation={navigation} />;

  // カレンダービュー
  const calendar = <TaskCalendarView tasks={tasks} />;

  // JSX
  return (
    <Container>
      <Segment>
        <Button 
          first 
          active={viewType === 'LIST'}
          onPress={() => setViewType('LIST')}
        >
          <Icon name="md-list" />
        </Button>
        <Button 
          last 
          active={viewType === 'CALENDAR'}
          onPress={() => setViewType('CALENDAR')}
        >
          <Icon name="md-calendar" />
        </Button>
      </Segment>
      <Content>
        { 
          viewType === 'LIST'
            ? taskList 
            : calendar
        }
      </Content>

      <Fab 
        position="bottomRight"
        active={false}
        direction=""
        containerStyle={{ }}
        large
        primary
        onPress={() => navigation.navigate('新規作成')}
      >
        <Icon name="ios-add" secondary />
      </Fab>
    </Container>
  );
}
