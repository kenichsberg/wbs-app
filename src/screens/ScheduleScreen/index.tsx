import * as React from 'react';
import 'react-native-gesture-handler';
import { Container, Segment, Content, View, Body, Right, Text, Button, List, ListItem, Separator, Icon, Fab } from 'native-base';
import moment from 'moment';
import { TaskListView } from '/screens/TaskListView';
import { TaskCalendarView } from '/screens/TaskCalendarView';
import { TaskListProps } from '/navigations/types.tsx';
import { Task } from '/screens/CreateTaskScreen';

export type PartialTask = Partial<Task>;

type ViewType = 'LIST' | 'CALENDAR';

export const ScheduleScreen: React.FC<TaskListProps> = ({ navigation, route }) => {

  // プロセスオブジェクトを保持するstate
  const [tasks, setTasks] = React.useState<Array<PartialTask>>([]);

  // ビュー切り替え用state
  const [viewType, setViewType] = React.useState<ViewType>('LIST');

  // 引数を格納する変数
  let param: PartialTask = {};

  // 引数を受け取った時の処理
  React.useEffect(() => {
    if (route.params?.task) {
      param = route.params.task;

      // idがない場合は採番
      param.id = param.id ?? tasks.length;
      // paramと同じidがある場合は取り除く
      const tasksUnique = tasks.filter(task => task.id !== param.id);
      
      setTasks([...tasksUnique, param]);
    }
  }, [route.params?.task]);


  // ビューの切り替え
  let view: JSX.Element; 
  switch (viewType) {
    case 'LIST':
      view = <TaskListView tasks={tasks} navigation={navigation} />;
      break;
    case 'CALENDAR':
      view = <TaskCalendarView tasks={tasks} />;
      break;
    default:
      view = <Text>error</Text>;
  }

  // JSX
  return (
    <Container>
      <Segment>
        <Button 
          first 
          active={viewType === 'LIST'}
          onPress={() => setViewType('LIST')}
          data-test="list-view-button"
        >
          <Icon name="md-list" />
        </Button>
        <Button 
          last 
          active={viewType === 'CALENDAR'}
          onPress={() => setViewType('CALENDAR')}
          data-test="calendar-view-button"
        >
          <Icon name="md-calendar" />
        </Button>
      </Segment>
      <Content>
        { view }
      </Content>

      <Fab 
        position="bottomRight"
        active={false}
        containerStyle={{ }}
        onPress={() => navigation.navigate('CreateTask')}
        data-test="fab"
      >
        <Icon name="ios-add" />
      </Fab>
    </Container>
  );
}
