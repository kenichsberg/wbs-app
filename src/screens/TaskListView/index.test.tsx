import * as React from 'react';
import 'react-native';
import 'jest-enzyme';
import { shallow, ShallowWrapper } from 'enzyme';
import * as renderer from 'react-test-renderer';

import { TaskListView } from './index';
import { PartialTask } from '/screens/ScheduleScreen/';


const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn()
  },
  ...props
});


const task1 = {
  id: 0,
  category: '開発',
  taskName: 'Hello World',
  //startDatetimePlanned: '2020/4/1 9:00',
  //endDatetimePlanned: '2020/4/2 18:00',
  startDatetimePlanned: "\"2020-06-28T08:00:03.030Z\"",
  endDatetimePlanned: "\"2020-06-29T08:00:03.030Z\"",
  startDatetimeResult: null,
  endDatetimeResult: null,
  selectedDocument: 4
};

const task2 = {
  id: 1,
  category: '開発',
  taskName: 'ログイン機能開発',
  //startDatetimePlanned: '2020/4/3 9:00',
  //endDatetimePlanned: '2020/4/3 18:00',
  startDatetimePlanned: "\"2020-06-29T08:00:03.030Z\"",
  endDatetimePlanned: "\"2020-06-30T08:00:03.030Z\"",
  startDatetimeResult: null,
  endDatetimeResult: null,
  selectedDocument: 4
};

const tasks: Array<PartialTask>  = [task1, task2];


describe('TaskListView', () => {
  let wrapper: ShallowWrapper;
  let props: any;

  it("should render correctly without params", () => {
    props = createTestProps({
      tasks: []
    });

    const tree = renderer.create(<TaskListView {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render correctly with params", () => {
    props = createTestProps({
      tasks: tasks
    });

    const tree = renderer.create(<TaskListView {...props}  />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should navigate to edit task screen", () => {
    props = createTestProps({
      tasks: tasks
    });

    const navigate = jest.spyOn(props.navigation, 'navigate');

    const simulateClick = async () => {
      await wrapper.find('[data-test="edit-button"]').simulate('click');
      expect(navigate).toBeCalledWith('CreateTask');
    };

    simulateClick();
  });
});
