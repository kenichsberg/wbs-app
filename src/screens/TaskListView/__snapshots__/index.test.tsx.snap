// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskListView should render correctly with params 1`] = `
<View
  style={Object {}}
>
  <View
    bordered={true}
    style={
      Object {
        "backgroundColor": "#F0EFF5",
        "borderBottomWidth": 0.5,
        "borderColor": "#c9c9c9",
        "borderTopWidth": 0.5,
        "flex": 1,
        "height": 35,
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingTop": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#777",
          "fontFamily": "System",
          "fontSize": 12,
        }
      }
      uppercase={false}
    >
      開発
    </Text>
  </View>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={null}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderBottomWidth": 0.5,
          "borderColor": "#c9c9c9",
          "flexDirection": "row",
          "marginLeft": 16,
          "paddingBottom": 13,
          "paddingRight": 16,
          "paddingTop": 13,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": null,
            "alignSelf": null,
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
              "marginLeft": 10,
              "marginRight": 10,
            }
          }
          uppercase={false}
        >
          Hello World
        </Text>
        <Text
          note={true}
          style={
            Object {
              "color": "#808080",
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "200",
              "marginLeft": 10,
              "marginRight": 10,
            }
          }
          uppercase={false}
        >
          予定：
          2020年06月28日 17:00 〜 2020年06月29日 17:00
        </Text>
        <Text
          note={true}
          style={
            Object {
              "color": "#808080",
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "200",
              "marginLeft": 10,
              "marginRight": 10,
            }
          }
          uppercase={false}
        >
          成果物: 
          4
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "alignSelf": "center",
            "flex": 0.28,
            "paddingBottom": null,
            "paddingLeft": null,
            "paddingRight": null,
            "paddingTop": null,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": null,
              "backgroundColor": "transparent",
              "borderBottomWidth": null,
              "borderColor": "#007aff",
              "borderLeftWidth": null,
              "borderRadius": 5,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "borderWidth": undefined,
              "elevation": 0,
              "flexDirection": "row",
              "height": 45,
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingBottom": 6,
              "paddingTop": 6,
              "shadowColor": null,
              "shadowOffset": null,
              "shadowOpacity": null,
              "shadowRadius": null,
            }
          }
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={null}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderBottomWidth": 0.5,
          "borderColor": "#c9c9c9",
          "flexDirection": "row",
          "marginLeft": 16,
          "paddingBottom": 13,
          "paddingRight": 16,
          "paddingTop": 13,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": null,
            "alignSelf": null,
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
              "marginLeft": 10,
              "marginRight": 10,
            }
          }
          uppercase={false}
        >
          ログイン機能開発
        </Text>
        <Text
          note={true}
          style={
            Object {
              "color": "#808080",
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "200",
              "marginLeft": 10,
              "marginRight": 10,
            }
          }
          uppercase={false}
        >
          予定：
          2020年06月29日 17:00 〜 2020年06月30日 17:00
        </Text>
        <Text
          note={true}
          style={
            Object {
              "color": "#808080",
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "200",
              "marginLeft": 10,
              "marginRight": 10,
            }
          }
          uppercase={false}
        >
          成果物: 
          4
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "alignSelf": "center",
            "flex": 0.28,
            "paddingBottom": null,
            "paddingLeft": null,
            "paddingRight": null,
            "paddingTop": null,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": null,
              "backgroundColor": "transparent",
              "borderBottomWidth": null,
              "borderColor": "#007aff",
              "borderLeftWidth": null,
              "borderRadius": 5,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "borderWidth": undefined,
              "elevation": 0,
              "flexDirection": "row",
              "height": 45,
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingBottom": 6,
              "paddingTop": 6,
              "shadowColor": null,
              "shadowOffset": null,
              "shadowOpacity": null,
              "shadowRadius": null,
            }
          }
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`TaskListView should render correctly without params 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomWidth": 0.5,
        "borderColor": "#c9c9c9",
        "flexDirection": "row",
        "marginLeft": 16,
        "paddingBottom": 13,
        "paddingRight": 16,
        "paddingTop": 13,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      uppercase={false}
    >
      タスクを追加してください
    </Text>
  </View>
</View>
`;
